src/main/java/com/mraof/minestuck/client/gui/DialogueScreen.java

```diff
@@ -1,79 +1,160 @@
 	private int yOffset;
 	
 	private List<FormattedCharSequence> messageLines;
-	private final List<Button> responseButtons = new ArrayList<>();
+	private final List<List<DialogueButton>> responseButtonPages = new ArrayList<>();
+	private int responsePage = 0;
+	
+	//todo replace buttons/pages with a scrollbar
+	private DialogueButton previousButton;
+	private DialogueButton nextButton;
+	
+	private int animationTick = 0;
 	
 	DialogueScreen(int dialogueId, Dialogue.DialogueData dialogueData)
 	{
 		super(Component.empty());
 		
 		this.dialogueId = dialogueId;
 		this.dialogueData = dialogueData;
+		
+		this.animationData = dialogueData.animationData();
 	}
 	
 	@Override
 	public void init()
 	{
-		//TODO static gui height/width may not make sense with customizable gui sizing
 		yOffset = (this.height / 2) - (GUI_HEIGHT / 2);
-		xOffset = (this.width / 2) - (GUI_WIDTH / 2);
+		xOffset = (this.width / 2) - (GUI_WIDTH / 2) - (animationData.spriteWidth() / 2) - animationData.xOffset();
 		
-		this.messageLines = font.split(this.dialogueData.message(), 210);
+		Function<Component, Component> messageStyler = MinestuckConfig.CLIENT.npcDialogueTextColors.get()
+				? Function.identity()
+				: message -> message.copy().withStyle(style -> style.withColor((TextColor) null));
+		
+		this.messageLines = this.dialogueData.messages().stream()
+				.map(messageStyler)
+				.flatMap(message -> font.split(message, GUI_WIDTH - 20).stream())
+				.toList();
 		
 		recreateResponseButtons();
 	}
 	
 	public void recreateResponseButtons()
 	{
-		responseButtons.forEach(this::removeWidget);
-		responseButtons.clear();
+		responseButtonPages.forEach(dialogueButtons -> dialogueButtons.forEach(this::removeWidget));
+		responseButtonPages.clear();
+		
+		int startY = yOffset + 22 + (DialogueButton.TEXT_SPACING * messageLines.size());
+		int cumulativeButtonHeight = 0;
+		List<DialogueButton> pageButtons = new ArrayList<>();
 		
-		int startY = yOffset + 20 + 9 * messageLines.size() + 2;
-		for(int i = 0; i < this.dialogueData.responses().size(); i++)
+		for(Dialogue.ResponseData data : this.dialogueData.responses())
 		{
-			Dialogue.ResponseData data = this.dialogueData.responses().get(i);
-			
-			ExtendedButton entryButton = new ExtendedButton(xOffset + 20, startY + 20 * i, 190, 14, data.message(),
+			DialogueButton entryButton = new DialogueButton(dialogueData.guiBackground(), true, xOffset + 16, startY + cumulativeButtonHeight, 190, DialogueButton.NORMAL_DEFAULT_HEIGHT, data.message(),
 					button -> clickResponse(data));
 			
+			int entryHeight = entryButton.trueHeight + BUTTON_GAP;
+			
+			cumulativeButtonHeight += entryHeight;
+			
 			data.conditionFailure().ifPresent(failure -> {
 				entryButton.setTooltip(Tooltip.create(conditionFailMessage(failure.causes())));
 				entryButton.active = false;
 			});
 			
-			responseButtons.add(addRenderableWidget(entryButton));
+			//TODO if messageLines is large, it could still push buttons over the edge of the gui
```

Comment:
Should we try and solve this too?

Reply:
I think it should be fine unless you have a good and quick solution